Quarkus REST calls protected constructor for QueryParam/PathParam and fails with 404

### Describe the bug

We currently switch from Quarkus 3.8.6 to 3.20.0. We now have the problem that Quarkus REST tries to call the <b>protected</b> constructor of a type, which is used as path or query parameter. The REST call then fails with a HTTP 404.

With Quarkus 3.8.6 it works fine.

### Example resource

```java
@Path("/parameters")
public class ParametersResource {

    public static final String URI_STRING_TYPE_PATH = "/stringTypePath";


    @Path(URI_STRING_TYPE_PATH + "/{stringValue}")
    @POST
    @Produces(MediaType.TEXT_PLAIN)
    public String stringTypePath(@PathParam("stringValue") final StringType stringValue) {
        return stringValue.getValue();
    }

}
```

### StringType

```java
public class StringType {

    private String value;


    protected StringType() {
    }


    protected StringType(final String value) {
        this.value = value;
    }


    public String getValue() {
        return value;
    }


    public static StringType fromString(final String s) {
        return new StringType(s);
    }


    @Override
    public String toString() {
        return value;
    }
}
```

When now having a REST client and a test, the call will fail with 404 and the log below (see section `Actual behavior`).

Additional hints:
* When having a parent POM, the problem is not reproducible. But this is a different problem and I will create another issue for this.
* In the [documentation](https://quarkus.io/guides/rest#parameter-mapping) it is not completely clear which priority the different variants have. So does a constructor have a higher priority than the `fromString` method? At least it looks like that, but it should be documented.

### Expected behavior

The protected constructor should not be called, when this is not possible. It should behave like with Quarkus 3.8.6, where I think it was possible to call the protected constructor. Or as an alternative, other methods like the `fromString` should be called. In general this should be documented.

### Actual behavior

Quarkus REST tries to call the protected constructor and fails with the following error (only visible with DEBUG output):

```
[ERROR] org.acme.ParametersResourceTest.testStringTypePath -- Time elapsed: 0.008 s <<< ERROR!
org.jboss.resteasy.reactive.ClientWebApplicationException: Received: 'Not Found, status code 404' when invoking REST Client method: 'org.acme.rest.ParametersClientApi#stringTypePath'
	at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext.unwrapException(RestClientRequestContext.java:205)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.handleException(AbstractResteasyReactiveContext.java:329)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:175)
	at io.smallrye.context.impl.wrappers.SlowContextualRunnable.run(SlowContextualRunnable.java:19)
	at org.jboss.resteasy.reactive.client.handlers.ClientSwitchToRequestContextRestHandler$1$1.handle(ClientSwitchToRequestContextRestHandler.java:38)
	at org.jboss.resteasy.reactive.client.handlers.ClientSwitchToRequestContextRestHandler$1$1.handle(ClientSwitchToRequestContextRestHandler.java:35)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:252)
	at io.vertx.core.impl.ContextInternal.lambda$runOnContext$0(ContextInternal.java:50)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	Suppressed: java.lang.IllegalAccessError: class org.base.types.StringType$quarkusrestparamConverter$ tried to access protected method 'void org.base.types.StringType.<init>(java.lang.String)' (org.base.types.StringType$quarkusrestparamConverter$ is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @718f805a; org.base.types.StringType is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @54d18072)
		at org.base.types.StringType$quarkusrestparamConverter$.convert(Unknown Source)
		at org.jboss.resteasy.reactive.server.handlers.ParameterHandler.handleResult(ParameterHandler.java:108)
		at org.jboss.resteasy.reactive.server.handlers.ParameterHandler.handle(ParameterHandler.java:59)
		at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:137)
		at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
		at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:638)
		at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
		at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
		at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
		at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
		at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
		at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
		... 2 more
	Suppressed: jakarta.ws.rs.NotFoundException: HTTP 404 Not Found
		at org.jboss.resteasy.reactive.server.handlers.ParameterHandler.handleResult(ParameterHandler.java:113)
		at org.jboss.resteasy.reactive.server.handlers.ParameterHandler.handle(ParameterHandler.java:59)
		at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:137)
		at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
		at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:638)
		at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
		at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
		at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
		at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
		at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
		at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
		... 2 more
	Caused by: [CIRCULAR REFERENCE: java.lang.IllegalAccessError: class org.base.types.StringType$quarkusrestparamConverter$ tried to access protected method 'void org.base.types.StringType.<init>(java.lang.String)' (org.base.types.StringType$quarkusrestparamConverter$ is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @718f805a; org.base.types.StringType is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @54d18072)]
Caused by: jakarta.ws.rs.WebApplicationException: Not Found, status code 404
	at io.quarkus.rest.client.reactive.runtime.DefaultMicroprofileRestClientExceptionMapper.toThrowable(DefaultMicroprofileRestClientExceptionMapper.java:19)
	at io.quarkus.rest.client.reactive.runtime.MicroProfileRestClientResponseFilter.filter(MicroProfileRestClientResponseFilter.java:54)
	at org.jboss.resteasy.reactive.client.handlers.ClientResponseFilterRestHandler.handle(ClientResponseFilterRestHandler.java:21)
	at org.jboss.resteasy.reactive.client.handlers.ClientResponseFilterRestHandler.handle(ClientResponseFilterRestHandler.java:10)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.invokeHandler(AbstractResteasyReactiveContext.java:231)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	... 14 more
```

### How to Reproduce?

Reproducer: https://github.com/timonzi/path-and-query-param (`main` branch)

Steps to reproduce the behavior:
1. Build `path-and-query-param-base` (`mvn clean install`)
2. Build `path-and-query-param-usage` (`mvn clean install`) => Tests will fail

See also the README.md

Switch to [Quarkus 3.8.6](https://github.com/timonzi/path-and-query-param/tree/Quarkus_3.8.6) (branch `Quarkus_3.8.6`) and the problem is not reproducible.

### Output of `uname -a` or `ver`

Linux nb-timonz 6.8.0-52-generic #53~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Jan 15 19:18:46 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

OpenJDK 64-Bit Server VM Temurin-21.0.5+11 (build 21.0.5+11-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.20.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6

### Additional information

_No response_

--- a/docs/src/main/asciidoc/rest.adoc
+++ b/docs/src/main/asciidoc/rest.adoc
@@ -2881,17 +2881,19 @@ However, you can change this default behavior and constrain a provider to:
 === Parameter mapping
 
 All <<request-parameters,Request Parameters>> can be declared as link:{jdkapi}/java/lang/String.html[`String`],
-but also any of the following types:
+but also as any of the following types:
 
-- Types for which a link:{jaxrsapi}/jakarta/ws/rs/ext/ParamConverter.html[`ParamConverter`] is available via a registered
+1. Types for which a link:{jaxrsapi}/jakarta/ws/rs/ext/ParamConverter.html[`ParamConverter`] is available via a registered
 link:{jaxrsapi}/jakarta/ws/rs/ext/ParamConverterProvider.html[`ParamConverterProvider`].
-- Primitive types.
-- Types that have a constructor that accepts a single link:{jdkapi}/java/lang/String.html[`String`] argument.
-- Types that have a static method named `valueOf` or `fromString` with a single link:{jdkapi}/java/lang/String.html[`String`] argument
+2. Primitive types.
+3. Types that have a constructor that accepts a single link:{jdkapi}/java/lang/String.html[`String`] argument.
+4. Types that have a static method named `valueOf` or `fromString` with a single link:{jdkapi}/java/lang/String.html[`String`] argument
 that return an instance of the type. If both methods are present then `valueOf` will be used unless
 the type is an `enum` in which case `fromString` will be used.
-- link:{jdkapi}/java/util/List.html[`List<T>`], link:{jdkapi}/java/util/Set.html[`Set<T>`], or
-link:{jdkapi}/java/util/SortedSet.html[`SortedSet<T>`], where `T` satisfies any above criterion.
+5. link:{jdkapi}/java/util/List.html[`List<T>`], link:{jdkapi}/java/util/Set.html[`Set<T>`], or
+link:{jdkapi}/java/util/SortedSet.html[`SortedSet<T>`], where `T` satisfies 1, 3 or 4 above.
+
+For 3 and 4, constructor instantiation is prioritesed above method instantiation. Public, protected, and package private constructors and methods are supported for instantiation.
 
 The following example illustrates all those possibilities:
 

